keras/trainers/data_adapters/data_adapter_utils.py

```diff
@@ -4,6 +4,8 @@
 from keras.api_export import keras_export
 from keras.utils import tree
 
+NUM_SAMPLES_FOR_TENSOR_SPEC = 4
```

Comment:
Why not just 2?

Reply:
Also, these are batches, not samples

Reply:
I don't know, I just figured the odds of detecting different shapes would be better with more examples (but it will never be 100% correct).

I will rename.

Also, I need to add unit tests.

Reply:
The 2 main use cases are for images and sequences. In both cases you have a very high likelihood that 2 consecutive batches will have different image sizes or sequence lengths if they are dynamic. So IMO 2 is good enough and minimizes overhead. Of course, it's not 100% accurate, as you say.